emergence_lib/src/simulation/pathfinding.rs

```diff
@@ -1,35 +1,75 @@
+#[derive(Resource)]
+pub struct PassabilityCache {
+    /// The [`MapFilter`] inner type which caches whether a given position is passable, and its
+    /// corresponding hex patch
+    inner: MapFilter,
+    /// Tile positions that were marked as impassable in the last
+    /// [`update_from_impassable_query`](PassabilityCache::update_from_impassable_query) operation
+    previously_impassable: HashSet<TilePos>,
 }
 
-impl PassableFilters {
-    /// Create from a [`Impassable`] query
-    pub fn from_impassable_query(
-        query: &Query<&TilePos, With<Impassable>>,
-        template: &MapPositions,
-    ) -> PassableFilters {
-        PassableFilters {
-            inner: MapFilter::new_with_default(
-                true,
-                template,
-                query.iter().map(|pos| (*pos, false)),
-            ),
+impl PassabilityCache {
+    /// Creates new [`PassabilityCache`]
+    pub fn new(template: &MapPositions) -> PassabilityCache {
+        PassabilityCache {
+            inner: MapFilter::new_with_default(true, template, [].into_iter()),
+            previously_impassable: HashSet::new(),
         }
     }
+
+    /// Get neighbors associated with a given tile position, if it exists in the cache
+    pub fn get_patch(&self, tile_pos: &TilePos) -> Option<&HexPatch<MapData<bool>>> {
+        self.inner.get_patch(tile_pos)
+    }
+
+    /// Update from an [`Impassable`] query
+    pub fn update_from_impassable_query(&mut self, impassable: &Query<&TilePos, With<Impassable>>) {
+        let currently_impassable = HashSet::from_iter(impassable.iter().copied());
+
+        let update_to_passable = self.previously_impassable.difference(&currently_impassable);
+        let update_to_impassable = currently_impassable.difference(&self.previously_impassable);
+
+        self.inner
+            .update(update_to_passable.map(|position| (*position, true)));
+        self.inner
+            .update(update_to_impassable.map(|position| (*position, false)));
+
+        self.previously_impassable = currently_impassable;
+    }
+
+    /// Update from newly impassable (`&Query<&TilePos, (With<Impassable>, Changed<Impassable>)>`)
+    /// and newly passable queries (`&Query<&TilePos, (Without<Impassable>, Changed<Impassable>)>`)
+    pub fn update_from_changed_passable_queries(
+        &mut self,
+        newly_impassable: &Query<&TilePos, (With<Impassable>, Changed<Impassable>)>,
+        newly_passable: &Query<&TilePos, (Without<Impassable>, Changed<Impassable>)>,
+    ) {
+        let newly_impassable = HashSet::from_iter(newly_impassable.iter().copied());
+        let newly_passable = HashSet::from_iter(newly_passable.iter().copied());
+        self.previously_impassable = self
+            .previously_impassable
+            .union(&newly_impassable)
+            .copied()
+            .collect::<HashSet<TilePos>>()
+            .difference(&newly_passable)
+            .copied()
+            .collect::<HashSet<TilePos>>();
```

Comment:
I think this could be "dangerous" for performance, as four `HashSets` are allocated in total.
I don't have an immediate idea on how to improve this, but there could be room for performance improvements here in the future

Reply:
Maybe something like this:

```rust
self.previously_impassible
    .iter()
    .chain(newly_impassible.iter())
    .filter(|elem| !newly_passable.contains(elem))
    .copied()
    .collect::<HashSet<TilePos>>();
```

This considers more elments so it has to do more processing, but only allocates one `HashSet`.
I'm not sure if this would be faster in the end, this would need to be benchmarked (I also didn't test my solution, so you might have to tweak it to get it to work).

Reply:
> ```rust
> self.previously_impassible
>     .iter()
>     .chain(newly_impassible.iter())
>     .filter(|elem| !newly_passable.contains(elem))
>     .copied()
>     .collect::<HashSet<TilePos>>();
> ```

I think something very similar to this is what we need:

```rust
self.previously_impassable = self
            .previously_impassable
            .iter()
            .filter(|position| !newly_passable.contains(position))
            .chain(newly_impassable.iter())
            .copied()
            .collect::<HashSet<TilePos>>();
```
            
One change here is that the `filter` happens on the `previously_impassable` elements alone, and then the `newly_impassable` elements are simply chained on.